Teuchos: Teuchos RCP Thread-Safety (#229)

MicheldeMessieres opened the pull request:
RCPNode was changed for thread safe reference counting.
Unit tests have been added for RCP, Array, ArrayView, ArrayRCP, Ptr and Tuple.

Added Trilinos_ENABLE_THREAD_SAFE - defaults OFF
Added Teuchos_ENABLE_THREAD_SAFE - defaults Trilinos_ENABLE_THREAD_SAFE

Both options will give a config error if turned on when
Trilinos_ENABLE_CXX11 is OFF.

This was tested on SEMS rebased to the following commit:

commit f924a1bf71316296d00ed1a198d949a7e421e28f
Merge: e28b90b 50b7a75
Author: Roscoe A. Bartlett <rabartl@sandia.gov>
Date:   Thu May 18 12:08:11 2017 -0600
    Merge branch 'develop' of github.com:trilinos/Trilinos into develop

SEMS results:

100% tests passed, 0 tests failed out of 2327

Label Time Summary:
Amesos               =  18.40 sec (14 tests)
Amesos2              =   9.97 sec (8 tests)
Anasazi              = 100.98 sec (71 tests)
AztecOO              =  16.10 sec (17 tests)
Belos                =  92.13 sec (64 tests)
Domi                 = 145.85 sec (125 tests)
Epetra               =  43.13 sec (61 tests)
EpetraExt            =  13.47 sec (10 tests)
FEI                  =  40.05 sec (43 tests)
Galeri               =   4.60 sec (9 tests)
GlobiPack            =   1.27 sec (6 tests)
Ifpack               =  57.36 sec (53 tests)
Ifpack2              =  43.66 sec (33 tests)
Intrepid             = 155.91 sec (152 tests)
Intrepid2            =  29.12 sec (73 tests)
Isorropia            =   7.71 sec (6 tests)
Kokkos               = 236.37 sec (21 tests)
KokkosKernels        =   5.20 sec (9 tests)
ML                   =  44.00 sec (34 tests)
MiniTensor           =   0.66 sec (2 tests)
MueLu                = 269.29 sec (56 tests)
NOX                  = 127.08 sec (101 tests)
OptiPack             =   5.92 sec (5 tests)
Panzer               = 282.71 sec (129 tests)
Phalanx              =   4.05 sec (18 tests)
Pike                 =   2.72 sec (7 tests)
Piro                 =  23.82 sec (12 tests)
ROL                  = 833.66 sec (133 tests)
RTOp                 =  13.55 sec (24 tests)
Rythmos              = 135.72 sec (83 tests)
SEACAS               =   8.57 sec (8 tests)
STK                  =  25.08 sec (11 tests)
Sacado               =  87.39 sec (292 tests)
Shards               =   1.11 sec (4 tests)
ShyLU                =   7.95 sec (5 tests)
Stokhos              =  89.51 sec (74 tests)
Stratimikos          =  28.77 sec (39 tests)
Teko                 = 157.35 sec (19 tests)
Tempus               = 496.08 sec (9 tests)
Teuchos              =  57.16 sec (127 tests)
ThreadPool           =  15.87 sec (10 tests)
Thyra                =  61.70 sec (80 tests)
Tpetra               = 132.67 sec (127 tests)
TrilinosCouplings    =  39.67 sec (19 tests)
Triutils             =   2.32 sec (2 tests)
Xpetra               =  41.30 sec (17 tests)
Zoltan               = 201.78 sec (16 tests)
Zoltan2              = 127.44 sec (97 tests)

Total Test time (real) = 1243.06 sec

ibaned commented:
If this is the way forward to fixing #1303, it would be great if it could be prioritized.

bartlettroscoe commented:
> If this is the way forward to fixing #1303, it would be great if it could be prioritized.

Need to do some more code review and more detailed testing and then it can be merged.  Have a few critical things to get done first like TriBITSPub/TriBITS#183 (which is holding up a lot of other stuff) then I can spend some time on this.

jwillenbring commented:
@bartlettroscoe Just wondering if you know when you might have a chance to review this.

bartlettroscoe commented:
> Just wondering if you know when you might have a chance to review this.

Hopefully in the next couple of weeks.

bartlettroscoe commented:
> Is there anything @MicheldeMessieres can do to help make this less work for you? For example, run additional testing? We are targeting wrapping this up by the end of September and want to leave sufficient time in case issues come up.

Sorry, just saw this (I am not getting many GitHub Notification emails).  I will do a trial merge and see how it looks.

bartlettroscoe commented:
I did a superficial review of the branch `thread-safe-rcp-229-wip`.  I just pushed the following commit to the branch `thread-safe-rcp-229-wip`:

```
74f30b4 "Clean up config options, how tests are enabled, etc. (#229)"
Author: Roscoe A. Bartlett <rabartl@sandia.gov>
Date:   Wed Aug 2 13:13:50 2017 -0600 (30 minutes ago)

M       cmake/CallbackSetupExtraOptions.cmake
M       cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake
M       packages/teuchos/CMakeLists.txt
M       packages/teuchos/core/cmake/Dependencies.cmake
M       packages/teuchos/core/cmake/TeuchosCore_config.h.in
M       packages/teuchos/core/cmake/Teuchos_config.h.in
M       packages/teuchos/core/test/MemoryManagement/CMakeLists.txt
M       packages/teuchos/core/test/MemoryManagement/General_MT_UnitTests.hpp
```

However, one of the multi-threaded unit tests ran in the ctest test `TeuchosCore_MemoryManagement_MT_UnitTests_MPI_1` failed showing:

```
12. Ptr_mtPtrDangling_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% 
Test FAILED because only 33 dangling references were detected but expected 40.

 countDanglingReferences == expectedDanglingReferences = 0 == true : FAILED ==> /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/teuchos/core/test/MemoryManagement/Ptr_MT_UnitTests_Decl.hpp:248
 unknownErrors = 0 == 0 : passed
 [FAILED]  (0.0362 sec) Ptr_mtPtrDangling_UnitTest
 Location: /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/teuchos/core/test/MemoryManagement/Ptr_MT_UnitTests_Decl.hpp:157
```

I submitted the Teuchos test results to:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3034474

After running the tests a few times, that one unit test in the tests `TeuchosCore_MemoryManagement_MT_UnitTests_MPI_1`  and `TeuchosCore_MemoryManagement_MT_Nightly_UnitTests_MPI_1` passes sometimes and fails other times.  That suggests that there may be a race condition of some type.  Since this test is related to catching dangling Ptr objects, that makes sense. The current implementation of that is not 100% safe or correct in multi-threaded code.  That will need to be fixed at some point in the future.

However, this failing test not need to stop the initial merge to the 'develop' branch because multi-threading will be disabled by default.

Next I will:
* Ensure no new warnings are generated ...
* Run Teuchos test suite with valgrind ...
* Do a coverage build with Teuchos and submit coverage results ...
* Run all Trilinos CI tests with and without these changes with thread-safety disabled and compare times ...
* Run all Trilinos CI tests with thread-safety turned on and compare runtimes ...
* Set up a nightly build of Teuchos for the multi-threaded tests (and submit to Specialized build on CDash and forward results to teuchos-regressions) ...

----

**DETAILED NOTES:**

(8/1/2017)

I merged the branch `thread-safe-rcp-229-wip` to Trilinos 'develop' with the commits being:

```
commit 61be098cfb001fb8511aeb3bb3ea6f4bca236030
Merge: ae3efdf 8766875
Author:     Roscoe A. Bartlett <rabartl@sandia.gov>
AuthorDate: Tue Aug 1 19:23:35 2017 -0600
Commit:     Roscoe A. Bartlett <rabartl@sandia.gov>
CommitDate: Tue Aug 1 19:23:35 2017 -0600

    Merge remote-tracking branch 'tech-xcorp/thread-safe-rcp-229-wip' into develop
    
    Pulls in thread-safe implementation of Teuchos Memory Management classes.
    Thread safety is disabled initially.


commit 8766875dccde6b438fa8a4cf799a40ce653d65c3
Author:     Author micheldemessieres <michel@demessieres.com>
AuthorDate: Mon Jan 25 17:54:33 2016 -0700
Commit:     micheldemessieres <michel@demessieres.com>
CommitDate: Mon Jun 19 12:13:33 2017 -0400

    Teuchos: Teuchos RCP Thread-Safety (#229)
    
    RCPNode was changed for thread safe reference counting.
    Unit tests have been added for RCP, Array, ArrayView, ArrayRCP, Ptr and Tuple.
    
    Added Trilinos_ENABLE_THREAD_SAFE - defaults OFF
    Added Teuchos_ENABLE_THREAD_SAFE - defaults Trilinos_ENABLE_THREAD_SAFE
    
    Both options will give a config error if turned on when
    Trilinos_ENABLE_CXX11 is OFF.
    
    Changed RCP performance tests to std::shared_ptr

M       cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake
M       packages/teuchos/CMakeLists.txt
M       packages/teuchos/core/cmake/Teuchos_config.h.in
M       packages/teuchos/core/src/Teuchos_Array.hpp
M       packages/teuchos/core/src/Teuchos_GlobalMPISession.cpp
M       packages/teuchos/core/src/Teuchos_GlobalMPISession.hpp
M       packages/teuchos/core/src/Teuchos_RCP.hpp
M       packages/teuchos/core/src/Teuchos_RCPDecl.hpp
M       packages/teuchos/core/src/Teuchos_RCPNode.cpp
M       packages/teuchos/core/src/Teuchos_RCPNode.hpp
A       packages/teuchos/core/test/MemoryManagement/ArrayRCP_MT_Nightly_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/ArrayRCP_MT_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/ArrayRCP_MT_UnitTests_Decl.hpp
A       packages/teuchos/core/test/MemoryManagement/ArrayView_MT_Nightly_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/ArrayView_MT_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/ArrayView_MT_UnitTests_Decl.hpp
A       packages/teuchos/core/test/MemoryManagement/Array_MT_Nightly_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/Array_MT_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/Array_MT_UnitTests_Decl.hpp
M       packages/teuchos/core/test/MemoryManagement/CMakeLists.txt
A       packages/teuchos/core/test/MemoryManagement/General_MT_UnitTests.hpp
A       packages/teuchos/core/test/MemoryManagement/Ptr_MT_Nightly_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/Ptr_MT_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/Ptr_MT_UnitTests_Decl.hpp
A       packages/teuchos/core/test/MemoryManagement/RCP_Abort.cpp
A       packages/teuchos/core/test/MemoryManagement/RCP_MT_Nightly_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/RCP_MT_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/RCP_MT_UnitTests_Decl.hpp
M       packages/teuchos/core/test/MemoryManagement/RCP_Performance_UnitTests.cpp
M       packages/teuchos/core/test/MemoryManagement/RCP_UnitTests.cpp
M       packages/teuchos/core/test/MemoryManagement/TestClasses.hpp
A       packages/teuchos/core/test/MemoryManagement/Tuple_MT_Nightly_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/Tuple_MT_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/Tuple_MT_UnitTests_Decl.hpp
M       packages/teuchos/parameterlist/src/Teuchos_ObjectBuilder.hpp
M       packages/teuchos/parameterlist/test/ObjectBuilder/ObjectBuilder_UnitTests.cpp
```

First, I noticed a change to a TriBITS file:

```
diff --git a/cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake b/cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake
index 42a2680..ee1eb0d 100644
--- a/cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake
+++ b/cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake
@@ -1880,6 +1880,16 @@ MACRO(TRIBITS_SETUP_ENV)
     ENDIF()
   ENDIF()
 
+  # Set up thread safety options
+  OPTION(${PROJECT_NAME}_ENABLE_THREAD_SAFE
+    "Enable thread safe code including RCP classes."
+    OFF )
+
+  # If thread safety was turned on and CXX11 is off give a config error
+  IF (${PROJECT_NAME}_ENABLE_THREAD_SAFE AND NOT ${PROJECT_NAME}_ENABLE_CXX11)
+    MESSAGE(FATAL_ERROR "You set Trilinos_ENABLE_THREAD_SAFE=ON, but Trilinos' support for CXX11 is not enabled (Trilinos_ENABLE_CXX11=OFF).  This is not allowed.  Please enable T
+  ENDIF ()
+
   # Set up some MPI info
 
   IF (TPL_ENABLE_MPI)
```

We don't want to add this to TriBITS (since TriBITS is not supposed to be specifically know about Trilinos packages).  I will move these options to where they should go.

I did a bunch of cleanup for the CMake configure options, how the multi-threaded tests are included, and am now testing the multi-threaded tests.

```
$ time ./do-configure -DTrilinos_ENABLE_THREAD_SAFE=ON \
  -DTrilinos_ENABLE_Teuchos=ON \
  &> configure.out 

real    0m10.664s
user    0m6.649s
sys     0m3.126s

$ make -j16

$ time ctest -j16

98% tests passed, 2 tests failed out of 128

Label Time Summary:
Teuchos    = 1021.78 sec (128 tests)

Total Test time (real) = 981.60 sec

The following tests FAILED:
	 14 - TeuchosCore_MemoryManagement_MT_UnitTests_MPI_1 (Failed)
	 15 - TeuchosCore_MemoryManagement_MT_Nightly_UnitTests_MPI_1 (Failed)

real    16m21.643s
user    30m5.067s
sys     2m30.982s
```

The failing unit tests in `TeuchosCore_MemoryManagement_MT_UnitTests_MPI_1` showed:

```
Running unit tests ...

0. Array_mtArrayMultipleReads_NonConst_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% [Passed] (0.101 sec)
1. Array_mtArrayMultipleReads_Const_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% [Passed] (0.0939 sec)
2. Array_mtArrayDanglingReference_NonConst_ReadValues_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% Range Errors: 0 Scrambles: 0 [Passed] (0.145 sec)
3. Array_mtArrayDanglingReference_Const_ReadValues_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% Range Errors: 0 Scrambles: 0 [Passed] (0.116 sec)
4. Array_mtArrayDanglingReference_NonConst_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% Danglers: 16 Test FAILED because it detected only 16 Dangling References but should have detected 20.
[Passed] (0.273 sec)
5. Array_mtArrayDanglingReference_Const_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% Danglers: 16 Test FAILED because it detected only 16 Dangling References but should have detected 20.
[Passed] (0.277 sec)
6. Array_mtArrayDanglingReferenceFirstCycle_NonConst_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% Danglers: 20 [Passed] (0.252 sec)
7. Array_mtArrayDanglingReferenceFirstCycle_Const_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% Danglers: 20 [Passed] (0.214 sec)
8. ArrayRCP_mtArrayRCPMultipleReads_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% [Passed] (0.0446 sec)
9. ArrayRCP_mtRCPofArrayRCPMultipleReads_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% [Passed] (0.0569 sec)
10. ArrayView_mtArrayViewMultipleReads_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% [Passed] (0.0716 sec)
11. ArrayView_mtArrayViewDangling_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% Danglers: 30 Scrambles: 0 [Passed] (0.0502 sec)
12. Ptr_mtPtrDangling_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% 
Test FAILED because only 33 dangling references were detected but expected 40.

 countDanglingReferences == expectedDanglingReferences = 0 == true : FAILED ==> /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/teuchos/core/test/MemoryManagement/Ptr_MT_UnitTests_Decl.hpp:248
 unknownErrors = 0 == 0 : passed
 [FAILED]  (0.0362 sec) Ptr_mtPtrDangling_UnitTest
 Location: /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/teuchos/core/test/MemoryManagement/Ptr_MT_UnitTests_Decl.hpp:157
 
13. RCP_mtRefCount_UnitTest ... [Passed] (0.00286 sec)
14. RCP_mtCreateIndependentRCP_UnitTest ... [Passed] (0.00547 sec)
15. RCP_mtTestGetExistingRCPNodeGivenLookupKey_UnitTest ... [Passed] (0.0572 sec)
16. RCP_mtRCPLastReleaseByAThread_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% [Passed] (0.00881 sec)
17. RCP_mtRCPLastReleaseByAThreadWithDealloc_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% [Passed] (0.00436 sec)
18. RCP_mtRCPLastReleaseByAThreadWithDeallocHandle_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% [Passed] (0.000658 sec)
19. RCP_mtRCPThreadCallsRelease_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% [Passed] (0.0007 sec)
20. RCP_mtRCPExtraData_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% [Passed] (0.000799 sec)
21. RCP_mtRCPWeakStrongDeleteRace_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% [Passed] (0.0043 sec)
22. RCP_mtRCPMixedWeakAndStrongConvertToStrong_UnitTest ... Weak converted with null 20 times and success 20 times. We want to see a mix of each. [Passed] (0.000739 sec)
23. Tuple_mtTupleMultipleReads_UnitTest ... Percent complete: 0% 11% 22% 33% 44% 55% 66% 77% 88% 100% [Passed] (0.0576 sec)

The following tests FAILED:
    12. Ptr_mtPtrDangling_UnitTest ... 

Total Time: 1.88 sec

Summary: total = 24, run = 24, passed = 23, failed = 1
```

The failing unit tests in `TeuchosCore_MemoryManagement_MT_Nightly_UnitTests_MPI_1` showed:

```
Running unit tests ...

0. Array_mtArrayMultipleReads_NonConst_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% [Passed] (72.8 sec)
1. Array_mtArrayMultipleReads_Const_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% [Passed] (73.4 sec)
2. Array_mtArrayDanglingReference_NonConst_ReadValues_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% Range Errors: 156 Scrambles: 7011 [Passed] (69.5 sec)
3. Array_mtArrayDanglingReference_Const_ReadValues_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% Range Errors: 176 Scrambles: 6836 [Passed] (69.7 sec)
4. Array_mtArrayDanglingReference_NonConst_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% Danglers: 19263 Test FAILED because it detected only 19263 Dangling References but should have detected 20000.
[Passed] (250 sec)
5. Array_mtArrayDanglingReference_Const_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% Danglers: 19539 Test FAILED because it detected only 19539 Dangling References but should have detected 20000.
[Passed] (185 sec)
6. Array_mtArrayDanglingReferenceFirstCycle_NonConst_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% Danglers: 20000 [Passed] (84.9 sec)
7. Array_mtArrayDanglingReferenceFirstCycle_Const_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% Danglers: 20000 [Passed] (85.8 sec)
8. ArrayRCP_mtArrayRCPMultipleReads_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% [Passed] (12.9 sec)
9. ArrayRCP_mtRCPofArrayRCPMultipleReads_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% [Passed] (16.6 sec)
10. ArrayView_mtArrayViewMultipleReads_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% [Passed] (30.5 sec)
11. ArrayView_mtArrayViewDangling_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% Danglers: 30000 Scrambles: 0 [Passed] (1.54 sec)
12. Ptr_mtPtrDangling_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% 
Test FAILED because only 39997 dangling references were detected but expected 40000.

 countDanglingReferences == expectedDanglingReferences = 0 == true : FAILED ==> /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/teuchos/core/test/MemoryManagement/Ptr_MT_UnitTests_Decl.hpp:248
 unknownErrors = 0 == 0 : passed
 [FAILED]  (0.831 sec) Ptr_mtPtrDangling_UnitTest
 Location: /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/teuchos/core/test/MemoryManagement/Ptr_MT_UnitTests_Decl.hpp:157
 
13. RCP_mtRefCount_UnitTest ... [Passed] (0.00101 sec)
14. RCP_mtCreateIndependentRCP_UnitTest ... [Passed] (0.0403 sec)
15. RCP_mtTestGetExistingRCPNodeGivenLookupKey_UnitTest ... [Passed] (0.0457 sec)
16. RCP_mtRCPLastReleaseByAThread_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% [Passed] (0.454 sec)
17. RCP_mtRCPLastReleaseByAThreadWithDealloc_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% [Passed] (0.467 sec)
18. RCP_mtRCPLastReleaseByAThreadWithDeallocHandle_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% [Passed] (0.464 sec)
19. RCP_mtRCPThreadCallsRelease_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% [Passed] (0.464 sec)
20. RCP_mtRCPExtraData_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% [Passed] (0.509 sec)
21. RCP_mtRCPWeakStrongDeleteRace_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% [Passed] (0.465 sec)
22. RCP_mtRCPMixedWeakAndStrongConvertToStrong_UnitTest ... Weak converted with null 11416 times and success 28584 times. We want to see a mix of each. [Passed] (0.469 sec)
23. Tuple_mtTupleMultipleReads_UnitTest ... Percent complete: 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% [Passed] (24.5 sec)

The following tests FAILED:
    12. Ptr_mtPtrDangling_UnitTest ... 

Total Time: 981 sec

Summary: total = 24, run = 24, passed = 23, failed = 1
```

Looks like something is wrong with the dangling reference test?

I submitted this to CDash using `make dashboard`:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3034474

Interestingly, the test `TeuchosCore_MemoryManagement_MT_UnitTests_MPI_1` did not fail that time.  This suggests that this might be a random failure. (The test `TeuchosCore_MemoryManagement_MT_Nightly_UnitTests_MPI_1` timed out at 10 minutes).

bartlettroscoe commented:
I completed the following review tasks:

* Check for compiler warnings: I did not find any warnings with GCC 4.8.4
* Check for valgrind warnings: I did not find any valgrind warnings that should not be there (i.e. there are tests that are designed to crash and leak memory)
* Check coverage: There are several missed lines of code but I did not see any major use case or significant lines of code that were not covered.

So far so good.

Lastly, I will do the following before I merge and push this branch:

* Run all Trilinos CI build tests with and without these changes with thread-safety disabled and compare times ...
* Run all Trilinos CI build tests with thread-safety turned on and compare runtimes to runtimes without these changes ...

----

**DETAILED NOTES:**

**Check for compiler warnings:**

(8/2/2017)

I did a fresh from-scratch build of Teuchos with thread safety disabled with:

```
$ time ./do-configure -DTrilinos_ENABLE_Teuchos=ON &> configure.out

real    0m10.018s
user    0m6.294s
sys     0m2.833s

$ time make -j16 &> make.out

real    2m48.841s
user    23m16.669s
sys     1m31.040s

$ time ctest -j16 &> ctest.out 

real    0m3.295s
user    0m7.298s
sys     0m9.451s
```

showing the disabled tests:

```
-- TeuchosCore_MemoryManagement_MT_UnitTests: NOT added test because EXCLUDE_IF_NOT_TRUE Teuchos_ENABLE_THREAD_SAFE_UNIT_TESTS='OFF'!
-- TeuchosCore_MemoryManagement_MT_Nightly_UnitTests: NOT added test because EXCLUDE_IF_NOT_TRUE Teuchos_ENABLE_THREAD_SAFE_UNIT_TESTS='OFF'!
-- TeuchosCore_RCP_PerformanceTests: NOT added test because Trilinos_TEST_CATEGORIES='NIGHTLY' does not match this test's CATEGORIES='PERFORMANCE'!
-- TeuchosCore_Array_PerformanceTests: NOT added test because Trilinos_TEST_CATEGORIES='NIGHTLY' does not match this test's CATEGORIES='PERFORMANCE'!
```

with test results:

```
100% tests passed, 0 tests failed out of 126

Label Time Summary:
Teuchos    =  36.74 sec (126 tests)

Total Test time (real) =   3.29 sec
```

Grepping for warnings shows:

```
$ grep "warning:" make.out | grep -v "/yaml-cpp/"
[emtpy]
```

[Done]

**Check for valgrind warnings:**

Now running valgrind:

```
$ time ctest -j16 -T memcheck &> ctest.memcheck.out

real    0m40.649s
user    4m32.211s
sys     0m26.394s
```

That produced:

```
MemCheck log files can be found here: ( * corresponds to test number)
/ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.4/MPI_RELEASE_DEBUG_SHARED_PT/Testing/Temporary/MemoryChecker.*.log
Memory checking results:
Memory Leak - 2062
Potential Memory Leak - 59173
Uninitialized Memory Read - 119
```

(8/3/2017)

These looks like to be almost all be coming from OpenMPI.  I could hunt down a valgrind suppression file for OpenMPI but I don't want to fight that battle.  Instead, what I need to do is to do a serial build and then run valgrind on that.  I did a serial build and I am getting:

```
MemCheck log files can be found here: ( * corresponds to test number)
/ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.4/SERIAL_RELEASE_DEBUG_SHARED_PT/Testing/Temporary/MemoryChecker.*.log
Memory checking results:
Memory Leak - 1
Potential Memory Leak - 113
Uninitialized Memory Read - 1
        Errors while running CTest
```

These all look to be of the form:

```
==118063== 128 bytes in 1 blocks are still reachable in loss record 1 of 1
==118063==    at 0x4A07C20: realloc (vg_replace_malloc.c:662)
==118063==    by 0x512C4E8: gomp_realloc (alloc.c:54)
==118063==    by 0x5138E7B: gomp_init_num_threads (proc.c:121)
==118063==    by 0x512D48D: initialize_env (env.c:1187)
==118063==    by 0x513B0B5: ??? (in /home/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/4.9.3/base/lib64/libgomp.so.1.0.0)
==118063==    by 0x512C002: ??? (in /home/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/4.9.3/base/lib64/libgomp.so.1.0.0)
==118063== 
```

I will go ahead and create a suppression file for that and run it again.

I came up with the valgrind suppressions file `suppressions-libgomp.txt`:

```
{
   libgomp_initialize_env
   Memcheck:Leak
   fun:realloc
   fun:gomp_realloc
   fun:gomp_init_num_threads
   fun:initialize_env
   obj:/home/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/4.9.3/base/lib64/libgomp.so.1.0.0
   obj:/home/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/4.9.3/base/lib64/libgomp.so.1.0.0
   obj:*
   obj:*
   obj:*
   obj:*
   obj:*
   obj:*
   obj:*
   obj:*
   obj:*
   obj:*
   obj:*
   obj:*
   obj:*
   obj:*
   obj:*
   obj:*
   obj:*
   obj:*
}

{
   libgomp_initialize_env_2
   Memcheck:Leak
   fun:realloc
   fun:gomp_realloc
   fun:gomp_init_num_threads
   fun:initialize_env
   obj:/home/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/4.9.3/base/lib64/libgomp.so.1.0.0
   obj:/home/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/4.9.3/base/lib64/libgomp.so.1.0.0
}
```

And using the `do-configure` script:

```
#!/bin/bash

cmake \
-DTPL_ENABLE_MPI=OFF \
-DCMAKE_BUILD_TYPE=RELEASE \
-DTrilinos_ENABLE_DEBUG=ON \
-DBUILD_SHARED_LIBS=ON \
-DTrilinos_ENABLE_SECONDARY_TESTED_CODE=OFF \
-DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/MpiReleaseDebugSharedPtSettings.cmake,cmake/std/BasicCiTestingSettings.cmake \
-DTrilinos_ENABLE_DEBUG_SYMBOLS=ON \
-DTrilinos_ENABLE_EXPLICIT_INSTANTIATION=ON \
-DTrilinos_ENABLE_CONFIGURE_TIMING=ON \
-DTrilinos_ENABLE_TESTS=ON \
-DTrilinos_TRACE_ADD_TEST=ON \
-DCTEST_BUILD_FLAGS=-j16 \
-DCTEST_PARALLEL_LEVEL=16 \
-DMEMORYCHECK_COMMAND_OPTIONS="-q --trace-children=yes --tool=memcheck \
  --leak-check=yes --workaround-gcc296-bugs=yes \
  --show-reachable=yes --num-callers=50 --gen-suppressions=all \
  --suppressions=$PWD/suppressions-libgomp.txt" \
"$@" \
../../../Trilinos
```

I did the configure, build, and valgrind testing with:

```
$ cd ~/Trilinos.base/BUILDS/GCC-4.8.4/SERIAL_RELEASE_DEBUG_SHARED_PT/

$ rm -r CMake*

$ time ./do-configure -DTrilinos_ENABLE_Teuchos=ON &> configure.out

real    0m9.106s
user    0m5.918s
sys     0m2.588s

$ make -j16 &> make.out

$ time ctest -j16 -T memcheck &> ctest.memcheck.out

real    0m27.531s
user    4m26.938s
sys     1m10.548s
```

This resulted in:

```

98% tests passed, 2 tests failed out of 117

Label Time Summary:
Teuchos    = 392.80 sec (117 tests)

Total Test time (real) =  27.50 sec

The following tests FAILED:
	 48 - TeuchosCore_TypeConversions_UnitTest (Failed)
	 95 - TeuchosComm_ParameterList_UnitTest_Parallel_one (Failed)
-- Processing memory checking output:
56/117 MemCheck: #112: TeuchosNumerics_DenseMatrix_example .......................................   Defects: 1
69/117 MemCheck: #14: TeuchosCore_MemoryManagement_RCP_Abort_Verify .............................   Defects: 28
76/117 MemCheck: #25: TeuchosCore_testDisablePrintActiveRcpNodesOnExit_print ....................   Defects: 1
105/117 MemCheck: #26: TeuchosCore_testDisablePrintActiveRcpNodesOnExit_noprint ..................   Defects: 1
MemCheck log files can be found here: ( * corresponds to test number)
/ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.4/SERIAL_RELEASE_DEBUG_SHARED_PT/Testing/Temporary/MemoryChecker.*.log
Memory checking results:
Memory Leak - 2
Potential Memory Leak - 28
Uninitialized Memory Read - 1
        Errors while running CTest
```

The test `TeuchosComm_ParameterList_UnitTest_Parallel` should not even run in a serial build I suspect.  The test `TeuchosCore_TypeConversions_UnitTest` does not fail when run normally (i.e. `ctest -j16`).

The defect reported in `TeuchosNumerics_DenseMatrix_example` is a long seen invalid read in DGEMM and is not related to the Teuchos Memory Management classes.

The defects reported in the tests `TeuchosCore_testDisablePrintActiveRcpNodesOnExit_print` and `TeuchosCore_testDisablePrintActiveRcpNodesOnExit_noprint` are known to lead memory (since that is what they are supposed to do) and they should not even be run with valgrind.

That brings us to the test `TeuchosCore_MemoryManagement_RCP_Abort_Verify` which shows 28 defects by valgrind.  That test causes the program to abort due to an exception being thrown in the destructor of the managed object.  Because of this abort, a bunch of memoyr is leaked.  Therefore, this test should not be run through valgrind either (other than to confirm that a bunch of memory is leaked when an abort occurs).

So I think the results of this valgrind testing looks pretty good.

[Done]

**Check coverage:**

Checking coverage by configuring with:

```
$ cd ~/Trilinos.base/BUILDS/GCC-4.8.4/MPI_RELEASE_DEBUG_SHARED_PT/
$ rm -r CMake*
$ time ./do-configure -DTrilinos_ENABLE_Teuchos=ON \
  -DTrilinos_ENABLE_COVERAGE_TESTING=ON \
  &> configure.out

real    0m10.754s
user    0m6.851s
sys     0m2.703s

$ time make dashboard &> make.dashboard.out

real    4m52.135s
user    30m45.431s
sys     2m27.186s
```

That submitted coverage results to:

* https://testing.sandia.gov/cdash/viewCoverage.php?buildid=3036176

In particular, we are interested in the coverage results under teuchos/core/src/

* https://testing.sandia.gov/cdash/viewCoverage.php?buildid=3036176&status=6&dir=packages/teuchos/core/src

and the files for the Teuchos Memory Management classes.

To focus on the Teuchos Memory Management classes, I ran the coverage queries:

* https://testing.sandia.gov/cdash/viewCoverage.php?buildid=3036176&status=6&dir=packages/teuchos/core/src&filtercount=1&showfilters=1&field1=filename/string&compare1=63&value1=RCP
* https://testing.sandia.gov/cdash/viewCoverage.php?buildid=3036176&status=6&dir=packages/teuchos/core/src&filtercount=1&showfilters=1&field1=filename/string&compare1=63&value1=ArrayView
* https://testing.sandia.gov/cdash/viewCoverage.php?buildid=3036176&status=6&dir=packages/teuchos/core/src&filtercount=1&showfilters=1&field1=filename/string&compare1=63&value1=Ptr
* 

The coverage data is as follows:

Filename | Status | Percentage | Lines not covered
-------- |:------:| ----------:| -----------------:
Teuchos_ArrayRCP.hpp | Satisfactory  | 98.43% | 6/382
Teuchos_ArrayRCPDecl.hpp | Complete  | 100.00% | 0/2
Teuchos_ArrayView.cpp | Medium  | 60.98% | 16/41
Teuchos_ArrayView.hpp | Satisfactory  | 95.06% | 8/162
Teuchos_ArrayViewDecl.hpp | Complete  | 100.00% | 0/8
Teuchos_Ptr.cpp | Complete  | 100.00% | 0/4
Teuchos_Ptr.hpp | Complete  | 100.00% | 0/30
Teuchos_PtrDecl.hpp | Complete  | 100.00% | 0/24
Teuchos_RCP.hpp | Satisfactory  | 99.49% | 1/195
Teuchos_RCPBoostSharedPtrConversions.hpp | Satisfactory  | 86.67% | 2/15
Teuchos_RCPBoostSharedPtrConversionsDecl.hpp | Satisfactory  | 60.00% | 4/10
Teuchos_RCPDecl.hpp | Complete  | 100.00% | 0/21
Teuchos_RCPNode.cpp | Satisfactory  | 95.19% | 9/187
Teuchos_RCPNode.hpp | Satisfactory  | 97.56% | 4/164
Teuchos_RCPStdSharedPtrConversions.hpp | Complete  | 100.00% | 0/15
Teuchos_RCPStdSharedPtrConversionsDecl.hpp | Complete  | 100.00% | 0/10
Teuchos_getBaseObjVoidPtr.hpp | Complete  | 100.00% | 0/2
Teuchos_getRawPtr.hpp | Complete  | 100.00% | 0/1

As you can see, the coverage is pretty good in most files but there are some significant holes.  Focusing on the missed lines covered the biggest problems are Teuchos::ArrayView, Teuchos::ArrayRCP.  The majority of missed lines of code for Teuchos::ArrayView are for template specializations that have nothing to do with this refactoring. There are some uncoverged lines that that can't really be run like the paraded default: statement for a switch() statement (in case new values get added and the switch statement does not get updated).  However, there are several cases that look like valid code that valid use cases would run which are not covered.  Many of these missed lines are for cases where an array object is zero length or some input object is null.  Some effort should be put in to cover some of these cases.  However, I looked over every missed line of code and I don't see any significant use case that has been missed.  Therefore, it is not urgent to cover these missed lines before we merge this branch.

[Done]

MicheldeMessieres commented:
@bartlettroscoe  Thanks for the updates - I’ve pushed a new commit for some fixes to the tests.
Since this is not part of the critical path I’m not sure if you want to concern with this now.

This commit implements 4 fixes:
- Resolved the broken dangling ptr test
 I had an internal time out on that test and that was not reliable since pressure on the threads may cause one to be held up.  I’ve changed this to just run until success (or the test times out for the ctest setting) which I think is a cleaner approach.  That also make this test consistent with design in the other tests.

- Removed a spurious macro you had commented on.

- Reduced nightly to run in about 1 minute so this should not time out anymore at the 10 minute setting.

- Fixed a warning for an unused method when debug is off and tests are running - I had missed this one earlier.

bartlettroscoe commented:
@MicheldeMessieres,

> I’ve pushed a new commit for some fixes to the tests.
Since this is not part of the critical path I’m not sure if you want to concern with this now.

I pulled the updated branch with this commit and tested with it.  Thanks for making these changes!  After the merge to the 'develop' branch later today, I plan to set up a nightly build with thread safety enabled so that we can maintain this feature going forward.  We will likely flip the switch to enable thread safety by default when move to Trilinos 13.0 (but could do so sooner since it does not really break any backward compatibility that I can see).

bartlettroscoe commented:
I ran the full Trilinos CI build test suite for Teuchos and downstream packages and there is no measurable increase in the test times for the new implementation with thread safety disabled.  And with thread safety enabled, the overall test times did not go up enough to even be statistically significant.  The overall testing time was in 10 sec out of over 620 sec in each case.  I would need to run a lot more test runs to get a better mean.

Based on all of this analysis, I think this looks good to go ahead and merge and push to 'develop', initially with thread safety disabled.  But I think we can pretty quickly turn thread safety on by default and see what happens.

----

**DETAILED NOTES:**

(8/3/2017)

**Compare CI build test times with and without these changes with thread-safety disabled:**

Now, comparing CI build test times starting from the Trilinos version on 'develop':

```
commit c464c5ea88cb3efeaa561ca1863632e460c60aa4
Author:     Jonathan Hu <jhu@sandia.gov>
AuthorDate: Thu Aug 3 14:01:38 2017 -0700
Commit:     Jonathan Hu <jhu@sandia.gov>
CommitDate: Thu Aug 3 16:51:15 2017 -0700

    Xpetra: suppress warnings for CUDA
    
    To address issues #1148 and #1133.
    
    Build/Test Cases Summary
    Enabled Packages: Xpetra
    Disabled Packages: PyTrilinos,Claps,TriKota
    Enabled all Forward Packages
    0) MPI_RELEASE_DEBUG_SHARED_PT => passed: passed=585,notpassed=0 (120.09 min)

M       packages/xpetra/src/BlockedMap/Xpetra_BlockedMap.hpp
M       packages/xpetra/src/BlockedMultiVector/Xpetra_BlockedMultiVector.hpp
M       packages/xpetra/src/BlockedVector/Xpetra_BlockedVector.hpp
M       packages/xpetra/src/Headers/Xpetra_ConfigDefs.hpp
M       packages/xpetra/src/Map/Xpetra_EpetraMap.hpp
M       packages/xpetra/src/MultiVector/Xpetra_EpetraMultiVector.hpp
M       packages/xpetra/src/Vector/Xpetra_EpetraIntVector.hpp
M       packages/xpetra/sup/Utils/Xpetra_IO.hpp
M       packages/xpetra/sup/Utils/Xpetra_MatrixMatrix.hpp
M       packages/xpetra/sup/Utils/Xpetra_ThyraUtils.hpp
```

First, running the CI build tests downstream from Teuchos with:

```
$ ./checkin-test-sems.sh --enable-all-packages=off --enable-packages=Teuchos --local-do-all
```

That returned:

```
FAILED: Trilinos/MPI_RELEASE_DEBUG_SHARED_PT: passed=2307,notpassed=1

Thu Aug  3 19:42:43 MDT 2017

Enabled Packages: Teuchos
Disabled Packages: PyTrilinos,Claps,TriKota Enabled all Forward Packages
Hostname: crf450.srn.sandia.gov
Source Dir: /home/rabartl/Trilinos.base/Trilinos/cmake/tribits/ci_support/../../..
Build Dir: /home/rabartl/Trilinos.base/BUILDS/CHECKIN/MPI_RELEASE_DEBUG_SHARED_PT

CMake Cache Varibles: -DTrilinos_TRIBITS_DIR:PATH=/home/rabartl/Trilinos.base/Trilinos/cmake/tribits -DTrilinos_ENABLE_TESTS:BOOL=ON -DTrilinos_TEST_CATEGORIES:STRING=BASIC -DTrilinos_ALLOW_NO_PACKAGES:BOOL=OFF -DDART_TESTING_TIMEOUT:STRING=300.0 -DBUILD_SHARED_LIBS=ON -DTrilinos_DISABLE_ENABLED_FORWARD_DEP_PACKAGES=ON -DTrilinos_TRACE_ADD_TEST=ON -DTrilinos_ENABLE_SECONDARY_TESTED_CODE:BOOL=OFF -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/MpiReleaseDebugSharedPtSettings.cmake,cmake/std/BasicCiTestingSettings.cmake -DTrilinos_ENABLE_Teuchos:BOOL=ON -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON -DTrilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES:BOOL=ON -DTrilinos_ENABLE_PyTrilinos:BOOL=OFF -DTrilinos_ENABLE_Claps:BOOL=OFF -DTrilinos_ENABLE_TriKota:BOOL=OFF
Make Options: -j16
CTest Options: -j16 

Pull: Not Performed
Configure: Passed (2.90 min)
Build: Passed (75.92 min)
Test: FAILED (10.62 min)

99% tests passed, 1 tests failed out of 2308

Label Time Summary:
Amesos               =  19.34 sec (14 tests)
Amesos2              =  10.68 sec (8 tests)
Anasazi              = 102.45 sec (71 tests)
AztecOO              =  16.36 sec (17 tests)
Belos                =  98.25 sec (70 tests)
Domi                 = 149.60 sec (125 tests)
Epetra               =  44.79 sec (61 tests)
EpetraExt            =  13.26 sec (10 tests)
FEI                  =  42.77 sec (43 tests)
Galeri               =   4.92 sec (9 tests)
GlobiPack            =   1.43 sec (6 tests)
Ifpack               =  58.97 sec (53 tests)
Ifpack2              =  35.25 sec (33 tests)
Intrepid             = 160.37 sec (152 tests)
Intrepid2            =  22.71 sec (82 tests)
Isorropia            =   7.58 sec (6 tests)
KokkosKernels        =   2.36 sec (9 tests)
ML                   =  44.25 sec (34 tests)
MiniTensor           =   0.34 sec (2 tests)
MueLu                = 233.46 sec (56 tests)
NOX                  = 140.58 sec (106 tests)
OptiPack             =   6.12 sec (5 tests)
Panzer               = 278.03 sec (132 tests)
Phalanx              =   4.72 sec (19 tests)
Pike                 =   2.56 sec (7 tests)
Piro                 =  26.53 sec (12 tests)
ROL                  = 757.83 sec (137 tests)
RTOp                 =   9.86 sec (24 tests)
Rythmos              = 154.31 sec (83 tests)
STK                  =  26.60 sec (11 tests)
Sacado               =  39.34 sec (292 tests)
Shards               =   0.78 sec (4 tests)
ShyLU                =   7.71 sec (5 tests)
Stokhos              =  87.49 sec (75 tests)
Stratimikos          =  31.40 sec (39 tests)
Teko                 =  78.79 sec (19 tests)
Tempus               = 693.34 sec (10 tests)
Teuchos              =  44.68 sec (126 tests)
Thyra                =  63.62 sec (80 tests)
Tpetra               = 136.24 sec (134 tests)
TrilinosCouplings    =  41.61 sec (19 tests)
Triutils             =   2.42 sec (2 tests)
Xpetra               =  40.74 sec (17 tests)
Zoltan2              = 132.91 sec (97 tests)

Total Test time (real) = 637.02 sec

The following tests FAILED:
	1384 - FEI_fei_ubase_MPI_3 (Failed)
Errors while running CTest

Total time for MPI_RELEASE_DEBUG_SHARED_PT = 89.44 min
```

(Hopefully that test `FEI_fei_ubase_MPI_3` is just a fluke.)

I then merged in the branch 'tech-xcorp/thread-safe-rcp-229-wip' which gives the updated commits:

```
1752cfa "Merge remote-tracking branch 'tech-xcorp/thread-safe-rcp-229-wip' into develop (#229)" <rabartl@sandia.gov> [Thu Aug 3 20:23:01 2017 -0600] (33 seconds ago)
b6c1696 "Teuchos: RCP Thread Safe Unit Test Updates (#229)" <michel@demessieres.com> [Thu Aug 3 16:04:39 2017 -0400] (6 hours ago)
M       packages/teuchos/core/test/MemoryManagement/ArrayRCP_MT_Nightly_UnitTests.cpp
M       packages/teuchos/core/test/MemoryManagement/ArrayRCP_MT_UnitTests_Decl.hpp
M       packages/teuchos/core/test/MemoryManagement/ArrayView_MT_Nightly_UnitTests.cpp
M       packages/teuchos/core/test/MemoryManagement/ArrayView_MT_UnitTests_Decl.hpp
M       packages/teuchos/core/test/MemoryManagement/Array_MT_Nightly_UnitTests.cpp
M       packages/teuchos/core/test/MemoryManagement/Array_MT_UnitTests_Decl.hpp
M       packages/teuchos/core/test/MemoryManagement/General_MT_UnitTests.hpp
M       packages/teuchos/core/test/MemoryManagement/Ptr_MT_Nightly_UnitTests.cpp
M       packages/teuchos/core/test/MemoryManagement/Ptr_MT_UnitTests_Decl.hpp
M       packages/teuchos/core/test/MemoryManagement/RCP_MT_Nightly_UnitTests.cpp
M       packages/teuchos/core/test/MemoryManagement/RCP_MT_UnitTests_Decl.hpp
M       packages/teuchos/core/test/MemoryManagement/Tuple_MT_Nightly_UnitTests.cpp
M       packages/teuchos/core/test/MemoryManagement/Tuple_MT_UnitTests_Decl.hpp

74f30b4 "Clean up config options, how tests are enabled, etc. (#229)" <rabartl@sandia.gov> [Wed Aug 2 13:13:50 2017 -0600] (31 hours ago)
M       cmake/CallbackSetupExtraOptions.cmake
M       cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake
M       packages/teuchos/CMakeLists.txt
M       packages/teuchos/core/cmake/Dependencies.cmake
M       packages/teuchos/core/cmake/TeuchosCore_config.h.in
M       packages/teuchos/core/cmake/Teuchos_config.h.in
M       packages/teuchos/core/test/MemoryManagement/CMakeLists.txt
M       packages/teuchos/core/test/MemoryManagement/General_MT_UnitTests.hpp

8766875 "Teuchos: Teuchos RCP Thread-Safety (#229)" <michel@demessieres.com> [Mon Jan 25 17:54:33 2016 -0700] (6 weeks ago)
M       cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake
M       packages/teuchos/CMakeLists.txt
M       packages/teuchos/core/cmake/Teuchos_config.h.in
M       packages/teuchos/core/src/Teuchos_Array.hpp
M       packages/teuchos/core/src/Teuchos_GlobalMPISession.cpp
M       packages/teuchos/core/src/Teuchos_GlobalMPISession.hpp
M       packages/teuchos/core/src/Teuchos_RCP.hpp
M       packages/teuchos/core/src/Teuchos_RCPDecl.hpp
M       packages/teuchos/core/src/Teuchos_RCPNode.cpp
M       packages/teuchos/core/src/Teuchos_RCPNode.hpp
A       packages/teuchos/core/test/MemoryManagement/ArrayRCP_MT_Nightly_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/ArrayRCP_MT_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/ArrayRCP_MT_UnitTests_Decl.hpp
A       packages/teuchos/core/test/MemoryManagement/ArrayView_MT_Nightly_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/ArrayView_MT_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/ArrayView_MT_UnitTests_Decl.hpp
A       packages/teuchos/core/test/MemoryManagement/Array_MT_Nightly_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/Array_MT_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/Array_MT_UnitTests_Decl.hpp
M       packages/teuchos/core/test/MemoryManagement/CMakeLists.txt
A       packages/teuchos/core/test/MemoryManagement/General_MT_UnitTests.hpp
A       packages/teuchos/core/test/MemoryManagement/Ptr_MT_Nightly_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/Ptr_MT_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/Ptr_MT_UnitTests_Decl.hpp
A       packages/teuchos/core/test/MemoryManagement/RCP_Abort.cpp
A       packages/teuchos/core/test/MemoryManagement/RCP_MT_Nightly_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/RCP_MT_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/RCP_MT_UnitTests_Decl.hpp
M       packages/teuchos/core/test/MemoryManagement/RCP_Performance_UnitTests.cpp
M       packages/teuchos/core/test/MemoryManagement/RCP_UnitTests.cpp
M       packages/teuchos/core/test/MemoryManagement/TestClasses.hpp
A       packages/teuchos/core/test/MemoryManagement/Tuple_MT_Nightly_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/Tuple_MT_UnitTests.cpp
A       packages/teuchos/core/test/MemoryManagement/Tuple_MT_UnitTests_Decl.hpp
M       packages/teuchos/parameterlist/src/Teuchos_ObjectBuilder.hpp
M       packages/teuchos/parameterlist/test/ObjectBuilder/ObjectBuilder_UnitTests.cpp
```

with the top commit:

```
commit 1752cfab0defdbf573f68fb618edb047a393245f
Merge: c464c5e b6c1696
Author:     Roscoe A. Bartlett <rabartl@sandia.gov>
AuthorDate: Thu Aug 3 20:23:01 2017 -0600
Commit:     Roscoe A. Bartlett <rabartl@sandia.gov>
CommitDate: Thu Aug 3 20:23:01 2017 -0600

    Merge remote-tracking branch 'tech-xcorp/thread-safe-rcp-229-wip' into develop (#229)
```

I then reran the CI build tests downstream from Teuchos with:

```
$ ./checkin-test-sems.sh --enable-all-packages=off --enable-packages=Teuchos --local-do-all
```

That gave the results:

```
passed: Trilinos/MPI_RELEASE_DEBUG_SHARED_PT: passed=2309,notpassed=0

Thu Aug  3 21:52:57 MDT 2017

Enabled Packages: Teuchos
Disabled Packages: PyTrilinos,Claps,TriKota Enabled all Forward Packages
Hostname: crf450.srn.sandia.gov
Source Dir: /home/rabartl/Trilinos.base/Trilinos/cmake/tribits/ci_support/../../..
Build Dir: /home/rabartl/Trilinos.base/BUILDS/CHECKIN/MPI_RELEASE_DEBUG_SHARED_PT

CMake Cache Varibles: -DTrilinos_TRIBITS_DIR:PATH=/home/rabartl/Trilinos.base/Trilinos/cmake/tribits -DTrilinos_ENABLE_TESTS:BOOL=ON -DTrilinos_TEST_CATEGORIES:STRING=BASIC -DTrilinos_ALLOW_NO_PACKAGES:BOOL=OFF -DDART_TESTING_TIMEOUT:STRING=300.0 -DBUILD_SHARED_LIBS=ON -DTrilinos_DISABLE_ENABLED_FORWARD_DEP_PACKAGES=ON -DTrilinos_TRACE_ADD_TEST=ON -DTrilinos_ENABLE_SECONDARY_TESTED_CODE:BOOL=OFF -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/MpiReleaseDebugSharedPtSettings.cmake,cmake/std/BasicCiTestingSettings.cmake -DTrilinos_ENABLE_Teuchos:BOOL=ON -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON -DTrilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES:BOOL=ON -DTrilinos_ENABLE_PyTrilinos:BOOL=OFF -DTrilinos_ENABLE_Claps:BOOL=OFF -DTrilinos_ENABLE_TriKota:BOOL=OFF
Make Options: -j16
CTest Options: -j16 

Pull: Not Performed
Configure: Passed (2.56 min)
Build: Passed (72.90 min)
Test: Passed (10.40 min)

100% tests passed, 0 tests failed out of 2309

Label Time Summary:
Amesos               =  20.06 sec (14 tests)
Amesos2              =  10.37 sec (8 tests)
Anasazi              = 102.67 sec (71 tests)
AztecOO              =  15.85 sec (17 tests)
Belos                =  97.01 sec (70 tests)
Domi                 = 147.49 sec (125 tests)
Epetra               =  41.27 sec (61 tests)
EpetraExt            =  13.21 sec (10 tests)
FEI                  =  41.22 sec (43 tests)
Galeri               =   4.15 sec (9 tests)
GlobiPack            =   1.10 sec (6 tests)
Ifpack               =  59.18 sec (53 tests)
Ifpack2              =  33.20 sec (33 tests)
Intrepid             = 156.55 sec (152 tests)
Intrepid2            =  19.26 sec (82 tests)
Isorropia            =   7.27 sec (6 tests)
KokkosKernels        =   2.19 sec (9 tests)
ML                   =  43.24 sec (34 tests)
MiniTensor           =   0.22 sec (2 tests)
MueLu                = 225.39 sec (56 tests)
NOX                  = 139.11 sec (106 tests)
OptiPack             =   6.01 sec (5 tests)
Panzer               = 273.62 sec (132 tests)
Phalanx              =   3.62 sec (19 tests)
Pike                 =   2.57 sec (7 tests)
Piro                 =  26.04 sec (12 tests)
ROL                  = 751.15 sec (137 tests)
RTOp                 =   9.44 sec (24 tests)
Rythmos              = 160.06 sec (83 tests)
STK                  =  25.13 sec (11 tests)
Sacado               =  33.60 sec (292 tests)
Shards               =   0.73 sec (4 tests)
ShyLU                =   8.09 sec (5 tests)
Stokhos              =  89.66 sec (75 tests)
Stratimikos          =  29.17 sec (39 tests)
Teko                 =  76.71 sec (19 tests)
Tempus               = 705.54 sec (10 tests)
Teuchos              =  39.57 sec (127 tests)
Thyra                =  56.47 sec (80 tests)
Tpetra               = 130.02 sec (134 tests)
TrilinosCouplings    =  41.41 sec (19 tests)
Triutils             =   2.18 sec (2 tests)
Xpetra               =  40.66 sec (17 tests)
Zoltan2              = 127.45 sec (97 tests)

Total Test time (real) = 624.09 sec

Total time for MPI_RELEASE_DEBUG_SHARED_PT = 85.87 min
```

Comparing, the overall test time went down from 637.02 to 624.09 sec.  That is not statistically significant but it is a good sign that performance of the refactored code with thread safety turned off is not a problem.

(8/4/2017)

Finally, I ran with thread safety turned on with:

```
$ ./checkin-test-sems.sh \
  --extra-cmake-options="-DTrilinos_ENABLE_THREAD_SAFE=ON" \
  --enable-all-packages=off --enable-packages=Teuchos \
  --local-do-all
```

and it provided the results:

```
passed: Trilinos/MPI_RELEASE_DEBUG_SHARED_PT: passed=2310,notpassed=0

Fri Aug  4 07:35:45 MDT 2017

Enabled Packages: Teuchos
Disabled Packages: PyTrilinos,Claps,TriKota Enabled all Forward Packages
Hostname: crf450.srn.sandia.gov
Source Dir: /home/rabartl/Trilinos.base/Trilinos/cmake/tribits/ci_support/../../..
Build Dir: /home/rabartl/Trilinos.base/BUILDS/CHECKIN/MPI_RELEASE_DEBUG_SHARED_PT

CMake Cache Varibles: -DTrilinos_ENABLE_THREAD_SAFE=ON -DTrilinos_TRIBITS_DIR:PATH=/home/rabartl/Trilinos.base/Trilinos/cmake/tribits -DTrilinos_ENABLE_TESTS:BOOL=ON -DTrilinos_TEST_CATEGORIES:STRING=BASIC -DTrilinos_ALLOW_NO_PACKAGES:BOOL=OFF -DDART_TESTING_TIMEOUT:STRING=300.0 -DBUILD_SHARED_LIBS=ON -DTrilinos_DISABLE_ENABLED_FORWARD_DEP_PACKAGES=ON -DTrilinos_TRACE_ADD_TEST=ON -DTrilinos_ENABLE_SECONDARY_TESTED_CODE:BOOL=OFF -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/MpiReleaseDebugSharedPtSettings.cmake,cmake/std/BasicCiTestingSettings.cmake -DTrilinos_ENABLE_Teuchos:BOOL=ON -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON -DTrilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES:BOOL=ON -DTrilinos_ENABLE_PyTrilinos:BOOL=OFF -DTrilinos_ENABLE_Claps:BOOL=OFF -DTrilinos_ENABLE_TriKota:BOOL=OFF
Extra CMake Options: -DTrilinos_ENABLE_THREAD_SAFE=ON Make Options: -j16 CTest Options: -j16 

Pull: Not Performed
Configure: Passed (2.39 min)
Build: Passed (73.18 min)
Test: Passed (10.58 min)

100% tests passed, 0 tests failed out of 2310

Label Time Summary:
Amesos               =  19.43 sec (14 tests)
Amesos2              =   9.57 sec (8 tests)
Anasazi              = 104.15 sec (71 tests)
AztecOO              =  15.92 sec (17 tests)
Belos                =  99.32 sec (70 tests)
Domi                 = 149.61 sec (125 tests)
Epetra               =  42.46 sec (61 tests)
EpetraExt            =  13.05 sec (10 tests)
FEI                  =  41.07 sec (43 tests)
Galeri               =   3.69 sec (9 tests)
GlobiPack            =   1.06 sec (6 tests)
Ifpack               =  56.91 sec (53 tests)
Ifpack2              =  34.74 sec (33 tests)
Intrepid             = 156.22 sec (152 tests)
Intrepid2            =  19.60 sec (82 tests)
Isorropia            =   7.90 sec (6 tests)
KokkosKernels        =   1.59 sec (9 tests)
ML                   =  43.05 sec (34 tests)
MiniTensor           =   0.28 sec (2 tests)
MueLu                = 227.58 sec (56 tests)
NOX                  = 142.16 sec (106 tests)
OptiPack             =   6.14 sec (5 tests)
Panzer               = 272.24 sec (132 tests)
Phalanx              =   3.02 sec (19 tests)
Pike                 =   2.51 sec (7 tests)
Piro                 =  25.30 sec (12 tests)
ROL                  = 766.92 sec (137 tests)
RTOp                 =  10.04 sec (24 tests)
Rythmos              = 154.69 sec (83 tests)
STK                  =  24.92 sec (11 tests)
Sacado               =  36.53 sec (292 tests)
Shards               =   0.58 sec (4 tests)
ShyLU                =   8.12 sec (5 tests)
Stokhos              =  91.89 sec (75 tests)
Stratimikos          =  32.85 sec (39 tests)
Teko                 =  87.85 sec (19 tests)
Tempus               = 716.02 sec (10 tests)
Teuchos              =  40.16 sec (128 tests)
Thyra                =  57.47 sec (80 tests)
Tpetra               = 132.99 sec (134 tests)
TrilinosCouplings    =  41.39 sec (19 tests)
Triutils             =   2.45 sec (2 tests)
Xpetra               =  40.83 sec (17 tests)
Zoltan2              = 128.02 sec (97 tests)

Total Test time (real) = 634.53 sec

Total time for MPI_RELEASE_DEBUG_SHARED_PT = 86.16 min
```

Wow, that is only a marginal increase in overall testing time from 624.09 to 634.53 sec.

The below table shows the summary of test times for the various cases where 'Orig' is the current implementation on the 'develop' branch, 'New noTS' is the new implementation with no thread safety enabled, and 'New TS' is the new implementation with thread safety turned on:

Package              | Orig   | New noTS | New TS
-------------------- | ------:|---------:| --------
Amesos               |  19.34 |    20.06 |   19.43
Amesos2              |  10.68 |    10.37 |    9.57
Anasazi              | 102.45 |   102.67 |  104.15
AztecOO              |  16.36 |    15.85 |   15.92
Belos                |  98.25 |    97.01 |   99.32
Domi                 | 149.60 |   147.49 |  149.61
Epetra               |  44.79 |    41.27 |   42.46
EpetraExt            |  13.26 |    13.21 |   13.05
FEI                  |  42.77 |    41.22 |   41.07
Galeri               |   4.92 |     4.15 |    3.69
GlobiPack            |   1.43 |     1.10 |    1.06
Ifpack               |  58.97 |    59.18 |   56.91
Ifpack2              |  35.25 |    33.20 |   34.74
Intrepid             | 160.37 |   156.55 |  156.22
Intrepid2            |  22.71 |    19.26 |   19.60
Isorropia            |   7.58 |     7.27 |    7.90
KokkosKernels        |   2.36 |     2.19 |    1.59
ML                   |  44.25 |    43.24 |   43.05
MiniTensor           |   0.34 |     0.22 |    0.28
MueLu                | 233.46 |   225.39 |  227.58
NOX                  | 140.58 |   139.11 |  142.16
OptiPack             |   6.12 |     6.01 |    6.14
Panzer               | 278.03 |   273.62 |  272.24
Phalanx              |   4.72 |     3.62 |    3.02
Pike                 |   2.56 |     2.57 |    2.51
Piro                 |  26.53 |    26.04 |   25.30
ROL                  | 757.83 |   751.15 |  766.92
RTOp                 |   9.86 |     9.44 |   10.04
Rythmos              | 154.31 |   160.06 |  154.69
STK                  |  26.60 |    25.13 |   24.92
Sacado               |  39.34 |    33.60 |   36.53
Shards               |   0.78 |     0.73 |    0.58
ShyLU                |   7.71 |     8.09 |    8.12
Stokhos              |  87.49 |    89.66 |   91.89
Stratimikos          |  31.40 |    29.17 |   32.85
Teko                 |  78.79 |    76.71 |   87.85
Tempus               | 693.34 |   705.54 |  716.02
Teuchos              |  44.68 |    39.57 |   40.16
Thyra                |  63.62 |    56.47 |   57.47
Tpetra               | 136.24 |   130.02 |  132.99
TrilinosCouplings    |  41.61 |    41.41 |   41.39
Triutils             |   2.42 |     2.18 |    2.45
Xpetra               |  40.74 |    40.66 |   40.83
Zoltan2              | 132.91 |   127.45 |  128.02
-------------------- | ------ | ---------- | ---------
**Total**            | **637.02** | **624.09** | **634.53**

Based on these timing results, I would not be apposed to turning on thread-safety by default since it has only a minimal impact on the overall test times.

However, we can poll the Trilinos developers and Trilinos users to see if they would like thread safety enabled by default or not.  But it would be nice to get some testing with real customer APPs to see the impact.

But with APPs more directly using Kokkos for basic array data-structures, the performance of Teuchos Arrays is likely going to be less of an issue as time goes on.  Thread safety and overall debug-mode checking are likely more important.

mhoemmen commented:
@bartlettroscoe and @MicheldeMessieres , this is awesome!  I'm really happy to see this all coming into fruition, and I appreciate your caution and careful testing.

> Based on these timing results, I would not be apposed to turning on thread-safety by default since it has only a minimal impact on the overall test times.

How do you think this will look for a debug build with `Teuchos_ENABLE_DEBUG:BOOL=ON`?  My guess is that thread safety won't cost much more than the debug checking.  Correct code before these changes should not have been updating reference counts in a thread-parallel region, so contention should not increase.  I'm not worried about modest slowdowns for the debug case, but just want to watch out for possible time-outs.  Thanks!

bartlettroscoe commented:
@mhoemmen 

> How do you think this will look for a debug build with Teuchos_ENABLE_DEBUG:BOOL=ON? My guess is that thread safety won't cost much more than the debug checking.

That was exactly the test case that I timed shown [above](https://github.com/trilinos/Trilinos/pull/1346#issuecomment-320260842).  I never timed the non-debug-mode case.  But @MicheldeMessieres did and you can see a detailed report of that in:

* https://github.com/trilinos/Trilinos/files/1012770/RCPReportP1.pdf
* https://github.com/trilinos/Trilinos/files/1081920/RCPReport.std.shared_ptr_bartlettroscoe.pdf

We would like some application codes to try turning on `-DTrilinos_ENABLE_THREAD_SAFE=ON` and see what happens.

I am about to add a release note to teuchos/ReleaseNotes.txt file that describes exactly the impact of these changes.  Basically, this is completely backward compatible except for for aborting if an exception is thrown in a destructor.  But since it is doubtful that any customers were depending on that behavior (because that is bad C++ form), all code that ran before will run after this change the same way.  Setting  `-DTrilinos_ENABLE_THREAD_SAFE=ON` only expands the set of allowable use cases.
